SLING-7699 add ResourceStream and unit tests

JEBailey opened the pull request:
updates java version to jdk8 as well. 
@rombert

rombert commented:
Thanks for the PR @JEBailey . Can you please open a separate email thread on dev@sling? I think additions to the API bundle warrant more scrutiny.

stefanseifert commented on pom.xml:
```diff
@@ -100,6 +100,13 @@
         </dependency>
+        <dependency>
+        	<groupId>org.apache.sling</groupId>
+        	<artifactId>org.apache.sling.testing.sling-mock</artifactId>
```
this introduces a cyclic dependency between the api artifact and the sling-mock artifact (also only in test scope). not sure if this will bring us trouble later.

JEBailey replied:
I'm not clear on how this could be an issue with the sling-mock being in the test scope

stefanseifert commented on pom.xml:
```diff
@@ -100,6 +100,13 @@
+        	<groupId>org.apache.sling</groupId>
+        	<artifactId>org.apache.sling.testing.sling-mock</artifactId>
+        	<version>2.2.12</version>
+        	<type>bundle</type>
```
please remove this type specifier - its not required.

JEBailey replied:
sure thing

JEBailey commented:
@stefanseifert  The object originally came from a different context so those additional convenience methods aren't as relevant anymore. I don't have a problem removing them. The only one that's problematic is the branch selector, which is also the one that gives the greatest performance improvement. 

As for adding it to the resource object. Something like:
resource.stream()
resource.streamChildren()
resource.stream(branchSelector)

or maybe 
resource.children()

JEBailey commented:
@stefanseifert added the changes to add this to the Resource object as default methods. I think that works a lot better.

ieugen commented:
I agree with @stefanseifert suggestions. Great work @JEBailey . Let's Sling to Java 8 and bring in the goodies :)

JEBailey commented:
@stefanseifert @ieugen some final comments. I updated the merge to focus solely on adding the stream to the resource object, as that works to well not to implement in that manner. I removed the streamChildren() method because:
1. there is already two methods to get the children
2. I'm not entirely convinced of the need to stream just the children versus what's there.
3. It can always be added later, which is better then trying to remove it

ieugen commented:
> I'm not entirely convinced of the need to stream just the children versus what's there

I think it has some merit when you wish to code in a functional style. If you can convert those children to a stream then I think it is ok to leave it out. 
An argument to leave it in is if we can make the underlying implementation so that when you have a large list of children it will fetch them in a paginated way behind the scene. Don't know if this is actually an issue or not. 

I also saw the version change and considering the move to Java 8 maybe we can jump to version 3.0 ?

JEBailey commented:
> I think it has some merit when you wish to code in a functional style. If you can convert those children to a stream then I think it is ok to leave it out.

You still have functional abilities with the current children methods

```
resource.listChildren().forEachRemaining(consumer);
resource.getChildren().forEach(consumer);
```

I like the idea around pagination. I have some some additional packages coming that might help with that.

As for the version jump. Right now that's just for resource api defined in package-info. If we're stating a java upgrade would require an API jump to the next level, would I need to do that for every package in the bundle. Or would this be a change to the bundle's versioning. I'm really unclear on that part. @rombert  thoughts?

JEBailey commented:
@stefanseifert I'm going to go ahead and update the version for the bundle to 3.0.0 since and leave the API at their respective levels.

stefanseifert commented:
one thought on naming the new methods: the name "stream" is not so intuitive - for sure it returns a stream, but the name of the method does not indicate that it returns the whole subtree (this is only mentioned in the comments). i think it is important to have very clear method names in the API.

esp. as one might look for a stream-based method for the direct children and uses this method without reading the javadocs. perhaps the method name does not even have to include "stream" word, as this is more a technical aspect and not the function of the API. if you compare it to java.util.List - there the stream() method returns the same view as the list, but as stream. this is not the case here.

concerning having a stream method for direct children: yes, we can add it later - but it may make sense to think about it now as well to ensure a good naming for all the stream methods. unfortunately the existing methods getChildren/listChildren cannot be used right-away to get a stream (only forEach/forEachRemaining).
